File: src/algorithm/leetcode/Problem142.java
package algorithm.leetcode;

public class Problem142 {
    public static ListNode detectCycle(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                ListNode p = head;
                ListNode q = fast;
                while (p != q) {
                    p = p.next;
                    q = q.next;
                }
                return q;
            }
        }
        return null;
    }

    public static void main(String[] args) {
        ListNode p1 = new ListNode(1);
        p1.next = p1;
        System.out.println(detectCycle(p1));
    }
}

File: src/algorithm_cpp/problem142.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == nullptr || head->next == nullptr) return nullptr;
	ListNode* fast = head->next->next;
	ListNode* slow = head->next;
	while (fast != slow) {
	    if (fast == nullptr || fast->next == nullptr) {
		return nullptr;
	    }
	    fast = fast->next->next;
	    slow = slow->next;
	}
	ListNode* ptr = head;
	while(ptr != slow) {
	    ptr = ptr->next;
	    slow = slow->next;
	}
	return ptr;
    }
};
